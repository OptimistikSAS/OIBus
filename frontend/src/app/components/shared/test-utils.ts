import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };
