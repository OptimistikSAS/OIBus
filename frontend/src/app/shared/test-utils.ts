import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';
import { Page } from './types';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };

export function toPage<T>(content: Array<T>, totalElements: number = content.length, num = 0, size = Math.max(20, totalElements)): Page<T> {
  return {
    content,
    totalElements,
    number: num,
    size,
    totalPages: Math.ceil(totalElements / size)
  };
}

export function emptyPage<T>(size = 20): Page<T> {
  return {
    totalElements: 0,
    content: [] as Array<T>,
    size,
    totalPages: 0,
    number: 0
  };
}
