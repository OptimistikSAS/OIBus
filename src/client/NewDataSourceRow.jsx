import React from 'react'
import PropTypes from 'prop-types'
import { Button, Form, FormGroup, Label, Input } from 'reactstrap'
import Select from './components/Select.jsx'
import apis from './services/apis'
import { AlertContext } from './context/AlertContext'

const NewDataSourceRow = ({ protocolList, addDataSource }) => {
  const [dataSource, setDataSource] = React.useState({ dataSourceId: '', enable: false, protocol: 'Modbus' })
  const { setAlert } = React.useContext(AlertContext)

  /**
   * Updates the data source's state
   * @param {*} event The change event
   * @returns {void}
   */
  const handleChange = (event) => {
    const { target } = event
    const { value } = target
    //  update the new data source's state
    setDataSource((prevState) => ({ ...prevState, [target.name]: value }))
  }

  /**
   * Submits a new data source
   * @returns {void}
   */
  const handleAddDataSource = async () => {
    if (dataSource.dataSourceId === '') return

    // Points is required on server side
    dataSource.points = []

    try {
      await apis.addSouth(dataSource)

      // add submitted dataSource to the table
      addDataSource(dataSource)
      // reset the line
      setDataSource({ dataSourceId: '', enable: false, protocol: 'Modbus' })
    } catch (error) {
      console.error(error)
      setAlert({ text: error.message, type: 'danger' })
    }
  }

  return (
    <Form className="oi-add-new">
      <FormGroup>
        <Label for="Id">
          New Data Source ID
        </Label>
        <Input value={dataSource.dataSourceId} id="Id" name="dataSourceId" type="text" onChange={handleChange} />
      </FormGroup>
      <FormGroup>
        <Label for="protocol">
          Protocol
        </Label>
        <Select value={dataSource.protocol} id="protocol" name="protocol" options={protocolList} onChange={handleChange} />
      </FormGroup>
      <FormGroup>
        <Button color="primary" onClick={() => handleAddDataSource()}>
          Add
        </Button>
      </FormGroup>
    </Form>
  )
}

NewDataSourceRow.propTypes = {
  protocolList: PropTypes.arrayOf(PropTypes.string).isRequired,
  addDataSource: PropTypes.func.isRequired,
}
export default NewDataSourceRow
