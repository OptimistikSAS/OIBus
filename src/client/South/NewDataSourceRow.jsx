import React from 'react'
import PropTypes from 'prop-types'
import { Button, Form, Col, Row } from 'reactstrap'
import { OIbText, OIbSelect } from '../components/OIbForm/index'

const NewDataSourceRow = ({ protocolList, addDataSource }) => {
  const [dataSourceId, setDataSourceId] = React.useState('')
  const [protocol, setProtocol] = React.useState(protocolList[0])
  /**
   * Updates the dataSource's state
   * @param {*} event The change event
   * @returns {void}
   */
  const handleAddDataSource = () => {
    //  update the new dataSource's state
    if (dataSourceId === '') return
    addDataSource({ dataSourceId, protocol })
  }

  const handleChange = (name, value) => {
    switch (name) {
      case 'dataSourceId':
        setDataSourceId(value)
        break
      case 'protocol':
      default:
        setProtocol(value)
        break
    }
  }
  return (
    <Form>
      <Row>
        <Col md="5">
          <OIbText
            label="New DataSource ID"
            value={dataSourceId}
            name="dataSourceId"
            onChange={handleChange}
            defaultValue=""
          />
        </Col>
        <Col md="3">
          <OIbSelect
            label="Protocol"
            value={protocol}
            name="protocol"
            options={protocolList}
            defaultValue={protocolList[0]}
            onChange={handleChange}
          />
        </Col>
        <Col md="3">
          <Button size="sm" className="oi-add-button" color="primary" onClick={() => handleAddDataSource()}>
            Add
          </Button>
        </Col>
      </Row>
    </Form>
  )
}

NewDataSourceRow.propTypes = {
  protocolList: PropTypes.arrayOf(PropTypes.string).isRequired,
  addDataSource: PropTypes.func.isRequired,
}
export default NewDataSourceRow
