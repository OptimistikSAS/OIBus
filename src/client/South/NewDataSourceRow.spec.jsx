import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'

import testConfig from '../../../tests/testConfig'
import NewDataSourceRow from './NewDataSourceRow.jsx'

let container
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

describe('NewDataSourceRow', () => {
  test('check NewDataSourceRow', () => {
    act(() => {
      ReactDOM.render(
        <NewDataSourceRow
          protocolList={testConfig.protocolList}
          addDataSource={() => (1)}
        />, container,
      )
    })
    expect(container).toMatchSnapshot()
  })
})
