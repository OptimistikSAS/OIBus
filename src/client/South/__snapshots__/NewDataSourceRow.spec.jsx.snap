// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewDataSourceRow check NewDataSourceRow 1`] = `
<div>
  <form
    class=""
  >
    <div
      class="row"
    >
      <div
        class="col-md-5"
      >
        <div
          class="form-group"
        >
          <label
            class=""
            for="dataSourceId"
          >
            New DataSource ID
          </label>
          <input
            class="oi-form-input form-control"
            id="dataSourceId"
            name="dataSourceId"
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
          />
        </div>
      </div>
      <div
        class="col-md-3"
      >
        <div
          class="form-group"
        >
          <label
            class=""
            for="protocol"
          >
            Protocol
          </label>
          <select
            class="oi-form-input form-control"
            id="protocol"
            name="protocol"
          >
            <option
              value="CSV"
            >
              CSV
            </option>
            <option
              value="OPCHDA"
            >
              OPCHDA
            </option>
            <option
              value="SQLDbToFile"
            >
              SQLDbToFile
            </option>
            <option
              value="FolderScanner"
            >
              FolderScanner
            </option>
            <option
              value="Modbus"
            >
              Modbus
            </option>
            <option
              value="OPCUA"
            >
              OPCUA
            </option>
            <option
              value="MQTT"
            >
              MQTT
            </option>
          </select>
          <div
            class="invalid-feedback"
          />
        </div>
      </div>
      <div
        class="col-md-3"
      >
        <button
          class="oi-add-button btn btn-primary btn-sm"
          type="button"
        >
          Add
        </button>
      </div>
    </div>
  </form>
</div>
`;
